samza-core/src/main/java/org/apache/samza/coordinator/communication/CoordinatorCommunicationContext.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.coordinator.communication;
+
+import org.apache.samza.config.Config;
+import org.apache.samza.metrics.MetricsRegistry;
+
+
+/**
+ * Contains components which can be used to build a {@link CoordinatorCommunication}.
+ * For example, provides access to job model and handling for worker states.
+ */
+public class CoordinatorCommunicationContext {
+  private final JobModelProvider jobModelProvider;
+  private final Config config;
+  private final MetricsRegistry metricsRegistry;
+
+  public CoordinatorCommunicationContext(JobModelProvider jobModelProvider, Config config,
+      MetricsRegistry metricsRegistry) {
+    this.config = config;
+    this.jobModelProvider = jobModelProvider;
+    this.metricsRegistry = metricsRegistry;
```

Comment:
Is the pattern of having a provider instead of direct job model because job model is absent? Or, is it tied to the fact that job model can change within the lifecycle and hence the provider.

If latter, what about config? Are we treating them as immutable within the lifecycle of the JC?

Reply:
1. This gives flexibility to change the job model within the lifecycle of the JC. In the current impl, job model won't change within the lifecycle of the JC though. 
2. It is easier to extend `JobModelProvider` to provide references to other objects. Now, we only need a reference to the serialized job model, but if other communication impls in the future need a reference to the deserialized job model, then it is easier to update the `JobModelProvider`.
3. It decouples the construction of the communication layer from the calculation of the job model.

Regarding config: My intention of the `config` in `CoordinatorCommunicationContext` was to be the config used to build the `CoordinatorCommunication`, and I guess I was thinking that part would be immutable (or more generally, any config that is used by the JC is immutable).`JobModelProvider` provides the `JobModel` which has the `Config` which is to be communicated to workers, and I suppose that might be mutable when job model changes. I see how it is confusing to have multiple `Config`s. I'll see if I can make this clearer.

Reply:
sure 👍